### ISSUE TITLE
Support for Pegasus Models?

### ISSUE THREAD

**Issue by eycheung** [NVIDIA/FasterTransformer#485]
### Branch/Tag/Commit

main

### Docker Image Version

n/a

### GPU name

n/a

### CUDA Driver

n/a

### Reproduced Steps

I'm wondering if there is any way to support Pegasus models (https://huggingface.co/docs/transformers/model_doc/pegasus#transformers.PegasusModel) with Fastertransformer in the near future? 

The architecture is very similar to BART models, except for the sinusoidal position embeddings. Wondering if it's straightforward to add a different position_embedding_type so we can also run Pegasus Models or if there is a simple workaround I can use with the existing BART model.

**Comment by byshiue** [comment]
It should be use absolute position type if BART and pass the position embedding table you want to use into model directly.

**Comment by eycheung** [comment]
Gotcha, thanks I'll give this a try!

**Comment by eycheung** [comment]
Hi @byshiue , sorry I have another question about this. I've tried your suggestion but my results in FasterTransformer don't match the outputs expected from running the HuggingFace model. I'm wondering if I have misconfigured something.

Changes I've made:
1. I've changed the `FTBartEncoderWeight` and `FTBartDecodingWeight` classes where I've replaced the weights for:
* `layernorm_embedding.weight` and its corresponding bias with `t = torch.empty((1, 1), dtype=torch_weight_dtype).contiguous().cuda()`
* I'm not confident on this part, but I was just copying the logic from how `layer_norm.weight` is skipped if it is not an mbart model. But I am doing this since our pegasus model does not use this layernorm embedding.
2. Config variables I've changed from the example: 
* `mbart = True`
* `activation_type="relu"` 
* `layernorm_type="pre_layernorm"`
* `q_scaling= 1 / sqrt(head_size)`

Here is the config I am using if it is helpful:
```
PegasusConfig {
  "activation_dropout": 0.1,
  "activation_function": "relu",
  "add_bias_logits": false,
  "add_final_layer_norm": true,
  "architectures": [
    "PegasusForConditionalGeneration"
  ],
  "attention_dropout": 0.1,
  "bos_token_id": 0,
  "classif_dropout": 0.0,
  "classifier_dropout": 0.0,
  "d_model": 1024,
  "decoder_attention_heads": 16,
  "decoder_ffn_dim": 4096,
  "decoder_layerdrop": 0.0,
  "decoder_layers": 16,
  "decoder_start_token_id": 0,
  "dropout": 0.1,
  "encoder_attention_heads": 16,
  "encoder_ffn_dim": 4096,
  "encoder_layerdrop": 0.0,
  "encoder_layers": 16,
  "eos_token_id": 1,
  "extra_pos_embeddings": 1,
  "force_bos_token_to_be_generated": false,
  "forced_eos_token_id": 1,
  "gradient_checkpointing": false,
  "id2label": {
    "0": "LABEL_0",
    "1": "LABEL_1",
    "2": "LABEL_2"
  },
  "init_std": 0.02,
  "is_encoder_decoder": true,
  "label2id": {
    "LABEL_0": 0,
    "LABEL_1": 1,
    "LABEL_2": 2
  },
  "length_penalty": 0.8,
  "max_length": 256,
  "max_position_embeddings": 1024,
  "model_type": "pegasus",
  "normalize_before": true,
  "normalize_embedding": false,
  "num_beams": 8,
  "num_hidden_layers": 16,
  "pad_token_id": 0,
  "scale_embedding": true,
  "static_position_embeddings": true,
  "task_specific_params": {
    ...
  },
  "torch_dtype": "float32",
  "transformers_version": "4.20.1",
  "use_cache": true,
  "vocab_size": 96104
}
```

Thank you for the help!

**Comment by symphonylyh** [comment]
Hi @eycheung , there are two suggestions on your changes:

1. In addition to the `torch.empty` changes in the Python utils scripts, please comment out https://github.com/NVIDIA/FasterTransformer/blob/main/src/fastertransformer/models/bart/BartEncoder.cc#L560-L567 and https://github.com/NVIDIA/FasterTransformer/blob/main/src/fastertransformer/models/bart/BartDecoding.cc#L505-L512 in the FT source code as well. Because the Layernorm kernels can flexibly handle "layernorm without bias", but not "layernorm without weight and bias" -- this means no layernorm, and should rather be commented out.
2. `q_scaling` may cause some confusion. In short, if your `q_scaling` is same as BART, then you should use `q_scaling = 1.0` because the `1/sqrt(head_size)` has been treated as default and baked into FT source code already: https://github.com/NVIDIA/FasterTransformer/blob/main/src/fastertransformer/layers/attention_layers/UnfusedAttentionLayer.cc#L190

**Comment by symphonylyh** [comment]
And please give information on whether your model is more similar to mBART or BART (I saw you set `mBART=True`). If mBART, you should check if Pegasus model applies the same scaling step to the input embedding: https://github.com/NVIDIA/FasterTransformer/blob/main/examples/pytorch/bart/utils/ft_encoder.py#L132. This is one of the diffs between BART and mBART

**Comment by eycheung** [comment]
Thank you @symphonylyh! After following your suggestions, the output of the encoder looks somewhat close now! I think some deviations are expected so this is probably close enough. And yes, I think the model I am using is closer to mBART and should apply that scaling step.

I am still having a bit of trouble with the decoder though, where the decoding terminates immediately.
```
{'output_ids': array([[[   1,    0,    0,    0,    0,    0,    0,    0,    0,    0,
             0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
             0,    0,    0,    0,    0,    0,    0,    0,    0,    0],
         [ 139,    1,    0,    0,    0,    0,    0,    0,    0,    0,
             0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
             0,    0,    0,    0,    0,    0,    0,    0,    0,    0],
         [ 198,    1,    0,    0,    0,    0,    0,    0,    0,    0,
             0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
             0,    0,    0,    0,    0,    0,    0,    0,    0,    0],
         [ 107,    1,    0,    0,    0,    0,    0,    0,    0,    0,
             0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
             0,    0,    0,    0,    0,    0,    0,    0,    0,    0],
         [ 151,    1,    0,    0,    0,    0,    0,    0,    0,    0,
             0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
             0,    0,    0,    0,    0,    0,    0,    0,    0,    0],
         [5309,    1,    0,    0,    0,    0,    0,    0,    0,    0,
             0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
             0,    0,    0,    0,    0,    0,    0,    0,    0,    0],
         [ 182,    1,    0,    0,    0,    0,    0,    0,    0,    0,
             0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
             0,    0,    0,    0,    0,    0,    0,    0,    0,    0],
         [ 110,    1,    0,    0,    0,    0,    0,    0,    0,    0,
             0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
             0,    0,    0,    0,    0,    0,    0,    0,    0,    0]]],
       dtype=int32),
 'sequence_lengths': array([[1, 2, 2, 2, 2, 2, 2, 2]], dtype=int32)}
```

This is mentioned as one of the differences in the HuggingFace Pegasus vs. BART implementation:
```
the model starts generating with pad_token_id (which has 0 token_embedding) as the prefix.
```
Do you know if specifying the start and end token like:
```python
start_id=model_hf.config.decoder_start_token_id,
end_id=model_hf.config.eos_token_id,
```
in the `FTBartDecoding` class would be sufficient to address this difference? Or do you have a guess on what other parts I may need to change?

Thanks again!

**Comment by symphonylyh** [comment]
@eycheung yes, start_id and end_id should follow the model requirements. BART/mBART is using 2 (eos token) as both start and end. Pegasus is different, as you mentioned

**Comment by eycheung** [comment]
I see, so then to confirm this *should* work without any changes needed to the python utils or the C++/CUDA code, but I probably have something misconfigured on my end? Is this right?

If so, then I can close this and keep debugging. But if you have any hints on how to debug this or places I can look please let me know!
